File: DotNet/Examples/Conversation/ConversationExampleServer/AppendServer.cs
﻿using System;
using Shastra.Hydra.ConversationExampleDto;
using Shastra.Hydra.Conversations;
using Shastra.Hydra.Messaging;

namespace Shastra.Hydra.ConversationExampleServer
{
    class AppendServer
    {
        private readonly Conversation<ConversationDto> _conversation;
        private readonly IDisposable _subscription;
        private string _suffix;

        public AppendServer(Conversation<ConversationDto> conversation)
        {
            _conversation = conversation;
            _subscription = conversation.SkipErrors().Subscribe(OnNext);
        }

        private void OnNext(AugmentedMessage<ConversationDto> message)
        {
            // Ignore invalid messages
            switch (message.Message.MessageType) {
                    case MessageTypes.Init:
                    _suffix = message.Message.Data;
                    _conversation.SendAsync(new ConversationDto { MessageType = MessageTypes.Ack });
                    break;
                case MessageTypes.Request:
                    _conversation.SendAsync(new ConversationDto { MessageType = MessageTypes.Response, Data = message.Message.Data + _suffix });
                    break;
                case MessageTypes.End:
                    _subscription.Dispose();
                    _conversation.Dispose();
                    break;
            }
        }

    }
}

File: Java/hydra-examples/src/main/java/uk/co/shastra/hydra/conversationexampleserver/AppendServer.java
package uk.co.shastra.hydra.conversationexampleserver;

import rx.Subscription;
import rx.util.functions.Action1;
import uk.co.shastra.hydra.conversationexampledto.*;
import uk.co.shastra.hydra.conversations.Conversation;

public class AppendServer {
    private final Conversation<ConversationDto> conversation;
    private final Subscription subscription;
    private String suffix;

    public AppendServer(Conversation<ConversationDto> conversation)
    {
        this.conversation = conversation;
        subscription = conversation.getObservable().subscribe(new Action1<ConversationDto>() {
			@Override public void call(ConversationDto message) { onNext(message); }
		});
    }

    private void onNext(ConversationDto message)
    {
        // Ignore invalid messages
		try {
	        switch (message.getMessageType()) {
	                case INIT:
	                suffix = message.getData();
	                ConversationDto dtoInit = new ConversationDto();
	                dtoInit.setMessageType(MessageTypes.ACK);
					conversation.send(dtoInit);
	                break;
	            case REQUEST:
	            	ConversationDto dtoReq = new ConversationDto();
	                dtoReq.setMessageType(MessageTypes.RESPONSE);
	                dtoReq.setData(message.getData() + suffix);
	                conversation.send(dtoReq);
	                break;
	            case END:
	                subscription.unsubscribe();
	                conversation.close();
	                break;
	            default:
	            	break;
	        }
		} catch (Exception e) {}
    }
}
